pkg/cdi/cache.go

```diff
@@ -288,6 +312,126 @@ func (c *Cache) WriteSpec(raw *cdi.Spec, name string) error {
 	return spec.Write(true)
 }
 
+// CreateTransientSpec creates a transient Spec file with the given content.
+// The file is put in the highest priority Spec directory. The Spec content
+// is validated before anything is written to the file system. The generated
+// file name is guaranteed to be unique among all transient Specs. The file
+// is always written as JSON. The path to the generated file is returned on
+// success. Otherwise an error is returned together with an empty path.
+//
+// If a non-empty optional tag is given, it is made a distinctive part of
+// the generated file name. This can be used to help identify Spec sources
+// in cases where multiple clients generate transient Specs with identical
+// vendor and class, simply by using a unique tag per client.
+//
+// A transient Spec is one whose lifecycle is tied to that of an external
+// entity or to events external to CDI and CDI clients, for instance the
+// creation and removal of a particular container. Generated transient Spec
+// files should eventually be removed using RemoveTransientSpec().
+func (c *Cache) CreateTransientSpec(raw *cdi.Spec, tag string) (path string, err error) {
```

Comment:
There are two problems with using this API instead of a deterministic name:
- the generated name must be stored somewhere
- creating the file must be idempotent to handle the case that the process gets killed right after creating the file and before doing anything further.

With a deterministic file name that's easy (just overwrite any existing file with that name, delete the file with the name if it still exists).

With this helper function the caller would first need to read all spec to check whether the content is already there, then only write another transient file if not.

I prefer defining a naming convention for transient files that all vendors are expected to follow.

Reply:
> There are two problems with using this API instead of a deterministic name:
> 
> * the generated name must be stored somewhere
> * creating the file must be idempotent to handle the case that the process gets killed right after creating the file and before doing anything further.
> 
> With a deterministic file name that's easy (just overwrite any existing file with that name, delete the file with the name if it still exists).
> 
> With this helper function the caller would first need to read all spec to check whether the content is already there, then only write another transient file if not.
> 
> I prefer defining a naming convention for transient files that all vendors are expected to follow.

Yes with this API a crash at that point becomes a real immediate problem, unless the restarted client generates the Spec file again with different names referencing the container-specific CDI devices... and we probably can't trust the driver to do that.

Then we could go with something like this:

```
    // GetSpecName generates vendor+class-scoped spec file names.
    // If a non-empty driver is given, it should uniquely identify the caller among multiple possible
    // CDI users on the same host generating Spec files for the same vendor/class. This function
    // is idempotent. It returns the same name if called with the same arguments.
    func GetSpecName(vendor, class, driver string) string {
        // ...
    }

    // GetTransientSpecName generates vendor+class-scoped transient spec file names.
    // If a non-empty driver is given, it should uniquely identify the caller among multiple possible
    // CDI users on the same host generating Spec files for the same vendor/class. Similarly, if a
    // non-empty transientID is given, it should uniquely identify an external entity to which the
    // Spec file corresponds to. This can be for instance the ID of a container for which the Spec
    // entries are exclusively generated. This function is not idempotent. It returns a unique name
    // on each invocation.
    func GetTransientSpecName(vendor, class, driver string) string {
        // ...
    }

    // Spec files are written with the already existing WriteSpec()...
```

This separates file name generation from Spec file writing and provides separate convenience functions for transient and other Spec file name generation. Transient file names are internally are generated unique per call. The client has now a chance to store the file name before writing the actual file (using WriteSpec()). If the DRA device driver/plugin crashes at the critical/problematic point, it can just reused the already generated name if it has one, or generate a name if it did not manage to store one yet.

I assume that the DRA device driver/plugin is already stateful so storing state is not a problem. If the driver is stateless or storing a file name is a prohibitively insurmountable problem, an alternative would be combining these to a single function

```
    // GetSpecName generates vendor+class-scoped spec file names.
    // If a non-empty driver is given, it should uniquely identify the caller among multiple possible
    // CDI users on the same host generating Spec files for the same vendor/class. Similarly, if a
    // non-empty transientID is given, it should uniquely identify an external entity to which the
    // Spec file corresponds to. This can be for instance the ID of a container for which the Spec
    // entries are exclusively generated.
    func GetSpecName(vendor, class, driver, transientID string) string {
        // ...
    }
```

In this case we can't really guarantee uniqueness which was the main point of these additions. We have to rely in the caller to do the right thing based on the documented usage of the API function.

Reply:
Currently example DRA driver is stateless and it could be the case for real drivers as well. We ended up using CDI file path /var/run/cdi/<driver name>-<resource claim UID>.json as example driver name(default: example-driver.cdi.k8s.io) contains vendor domain in our case, so using both vendor and driver would look tautological.

Reply:
From other point of view it's not a big deal for the driver to refresh cdi cache on its start, which should give the driver all information it needs.

Reply:
@pohly > With this helper function the caller would first need to read all spec to check whether the content is already there, then only write another transient file if not.

This is done by cdi API as well as far as I know.

Reply:
@klueska your input here would be appreciated as you've written first(and the only existing for now) real DRA plugin.

Reply:
A helper function like GetSpecName would be useful. "vendor" and "class" seem a bit redundant when "driver" alone is already required to be unique, which is the case, right?

Reply:
No, in the scheme I sketched up the driver was an optional(ly non-empty) parameter. It is supposed to be unique only within the same vendor/class domain. It's there to allow multiple drivers that handle devices with identical vendor and class to coexist on the same node, yet use this function to generate non-conflicting file names.

For instance, if a vendor had one driver dealing with legacy HW and another dealing with recent ones, then the legacy could use "legacy" as the driver and the other could use "" (or "current" or something equivalent).

Reply:
> @pohly > With this helper function the caller would first need to read all spec to check whether the content is already there, then only write another transient file if not.
> 
> This is done by the CDI API as far as I know.

@bart0sh Do you mean that your example driver is doing this on the client side ?

Reply:
Current implementation of the example driver just creates a spec, marshals it into json and writes to /var/run/cdi/<driver name>-<resource claim UID>.json
However, it can use CDI cache API on the start to get all CDI files parsed, stored in the cache and monitored I guess.

Reply:
> Current implementation of the example driver just creates a spec, marshals it into json and writes to `/var/run/cdi/<driver name>-<resource claim UID>.json` However, it can use CDI cache API on the start to get all CDI files parsed, stored in the cache and monitored I guess.

Ah, okay... you meant that. I think we'd like to avoid such extra complexity in the driver, ideally keeping it as simple as `generate Spec content + GetSpecName() + WriteSpec()`. So IOW keeping it like you have it now, just using CDI-provided convenience functions to generate the Spec name (according to our preferred convention) and to write the content (making sure the changes are atomic wrt. Refresh()). Then the example driver would serve as a properly written template for driver writers to copy and modify.

Reply:
> One thing to keep in mind regarding the following:
> 
> > Then the example driver would serve as a properly written template for driver writers to copy and modify.
> 
> Writing transient spec files like this is likely _not_ going to be done by many (if any) actual driver implementors. I see it as the exception rather than the common case. Most people will write one spec file once (at machine bootup or at driver start) and then refer to the various devices defined within it as allocation requests come in.
> 
> The whole idea of generating specs on the fly was new to me when I looked at the example driver, and it is definitely _not_ something we plan to do in the NVIDIA driver.

If that is true, then wouldn't it make better sense to provide *separate* functions for transient and 'ordinary' Spec file name generation ? For instance `GetSpecName()` and  `GetTransientSpecName()`. Then `driver` and `transientID` would only be passed to the transient one, so most people would not need to even wonder about what they are and how to correctly pick them.

Reply:
> I think we'd like to avoid such extra complexity in the driver, ideally keeping it as simple as generate Spec content + GetSpecName() + WriteSpec()

Agreed.

> Writing transient spec files like this is likely not going to be done by many (if any) actual driver implementors.

That depends. If the driver needs some actual input from the user before preparing the hardware (like programming an FPGA) and that then affects how the hardware is presented in the container, then the CDI file has to be generated.

> GenerateSpecName() and GenerateTransientSpecName()

:+1:

Reply:
@klihub > generate Spec content + GetSpecName() + WriteSpec()

 I'm wondering if user even need to know the CDI file name? It should be possible to get it when CDI spec is created, but getting the name before writing the spec to the file might not be needed at all. Would it be enough in most cases to generate Spec content (spec := NewSpec(...))  + spec.Write() ?

Reply:
> I'm wondering if user even need to know the CDI file name?

Yes, for (debug) log messages.

Reply:
We also need the name for removing the file.

Reply:
> Yes, for (debug) log messages

spec.GetPath() should be enough for that.

Reply:
What I was trying to say is that user of the spec API might not be interested in knowing the file name at all. What's important for user is to know that spec is stored somewhere and the runtime can access it.

Reply:
> @klihub > generate Spec content + GetSpecName() + WriteSpec()
> 
> I'm wondering if user even need to know the CDI file name? It should be possible to get it when CDI spec is created (spec.GetPath()), but getting the name before writing the spec to the file might not be needed at all. Would it be enough in most cases to generate Spec content (spec := NewSpec(...)) + writing the spec (spec.Write()) ?

Hmm... if we this route that @pohly and @klueska seem to prefer then we'll need to add either `RemoveSpec(vendor, class), RemoveTransientSpec(vendor, class, driver, transientID)` or just a `RemoveSpec(name)`, because the generated name is not a path just a basename. It is `WriteSpec()` which picks the last spec directory and writes the Spec there, using the given name, also chosing a default encoding if necessary.

There would be other ways to do that (look up the Spec using the unique device name, get the path of that spec, the os.Remove() it). But if there is implicit logic/assumptions for Spec creation in the package, then it should rather provide functions with the same implicit knowledge/assumption for Spec removal.

Reply:
Correct, but not only that. CDI should be able to also understand if a spec really needs to be written, i.e. it should compare desired spec content with the content already present in the file/cache. If content is not changed, we should avoid writing it again. This can help avoiding races between driver writing the spec and runtime reading it at the same time.

This may be an overcomplication on my side and simple atomic writing should be enough. As far as I can see in [the code](https://github.com/container-orchestrated-devices/container-device-interface/blob/main/pkg/cdi/spec.go#L154) spec.Write() writes CDI file atomically.

Reply:
> > Writing transient spec files like this is likely not going to be done by many (if any) actual driver implementors.
> 
> That depends. If the driver needs some actual input from the user before preparing the hardware (like programming an FPGA) and that then affects how the hardware is presented in the container, then the CDI file has to be generated.

Our plan for this is to have separate „devices“ predeclared for these dynamic options that declare what needs to be injected base on configuring then device this way. This won’t work for all use cases but it will work for us.

Reply:
> Correct, but not only that. CDI should be able to also understand if a spec really needs to be written, i.e. it should compare desired spec content with the content already present in the file/cache. If content is not changed, we should avoid writing it again.

I'm not sure if any of the golang JSON encoders are stable (producing identical output for semantically identical input). I recall the stock one is not. So let's leave this as a potential future exercise/improvement.

Reply:
works for me